modules/editor/grammar/post.pegjs

```diff
@@ -0,0 +1,95 @@
+Document
+  = WP_Block_List
+
+WP_Block_List
+  = WP_Block*
+
+WP_Block
+  = WP_Block_Balanced
+  / WP_Block_Html
+
+WP_Block_Balanced
+  = s:WP_Block_Start ts:(!WP_Block_End c:Any { return c })+ e:WP_Block_End
+  { return {
+    blockType: s.blockType,
+    attrs: s.attrs,
+    rawContent: ts.join( '' ),
+  } }
+
+WP_Block_Html
+  = ts:(!WP_Block_Balanced c:Any { return c })+
+  {
+    return {
+      blockType: 'html',
+      attrs: {},
+      rawContent: ts.join('')
+    }
+  }
+
+WP_Block_Start
+  = "<!--" __ "wp:" blockType:WP_Block_Type attrs:WP_Block_Attribute_List _? "-->"
+  { return {
+    type: 'WP_Block_Start',
+    blockType,
+    attrs,
+    text: text()
+  } }
+
+WP_Block_End
+  = "<!--" __ "/wp" __ "-->"
```

Comment:
One of the purposes of these is to remain as out of the way as possible. A few people have asked me to not even have closing tags ;)

That said, I'm ok with either way now, and I'd expect us to refine the approach as we test with complex, long, and broken posts. The key is to balance readability of the HTML with the presence of structural hints.